ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/Util/HardFork/OracularClock.hs

```diff
@@ -0,0 +1,242 @@
+    OracularClock(..)
+  , new
+  , onSlotChange
+  , withinEachSlot
+  , withinTheCurrentSlot
+  ) where
+
+import           Control.Monad (replicateM_, void)
+import           Data.Foldable (toList)
+import           Data.Time
+import           GHC.Stack
+
+import           Ouroboros.Consensus.Util.IOLike
+import           Ouroboros.Consensus.Util.ResourceRegistry
+
+import           Cardano.Slotting.Slot
+
+import qualified Ouroboros.Consensus.BlockchainTime as BTime
+import           Ouroboros.Consensus.Util.STM (onEachChange)
+import           Ouroboros.Consensus.Util.Time (nominalDelay)
+
+import           Test.Util.HardFork.Future (Future, futureSlotLengths,
+                     futureTimeToSlot)
+import           Test.Util.Slots (NumSlots (..))
+import           Test.Util.Stream
+
+-- | A clock that knows the future
+--
+-- This clock's closure contains a 'BTime.SystemTime', a 'Future', and a
+-- 'NumSlots'. As the system time advances, the clock maintains a notion of the
+-- current 'SlotNo'. Once all 'NumSlots' have passed, the clock is /exhausted/
+-- and most of its methods block indefinitely from then on.
+--
+-- Most notably, 'waitUntilDone' instead blocks until the the clock is
+-- exhausted; so the continuation of that call should expect other threads
+-- using the rest of this clock's methods to be stuck (and eg reap them).
+--
+-- Note: though the wallclock-slot correspondence depends on the ledger state,
+-- we have designed our ledgers so that all nodes necessarily use the same
+-- correspondence in the absence of a Common Prefix violation. So this clock
+-- can indeed know /the/ future.
+data OracularClock m = OracularClock
+    { -- | Returns 'False' if the requested slot is already over
+      --
+      -- Note: blocks indefinitely if clock becomes exhausted before requested
+      -- slot.
+      blockUntilSlot :: SlotNo -> m Bool
+
+      -- | The current delay duration until the onset of the next slot
+      --
+      -- Note: blocks indefinitely if the clock is already exhausted.
+    , delayUntilNextSlot :: m NominalDiffTime
+
+      -- | A mock system time
+      --
+      -- Note: this 'BTime.systemTimeCurrent' blocks indefinitely if clock is
+      -- already exhausted.
+    , finiteSystemTime :: BTime.SystemTime m
+
+      -- | Convert 'Control.Monad.Class.MonadTime.getCurrentTime' to 'SlotNo'
+      --
+      -- INVARIANT: In @io-sim@, the @getCurrentTime@ use (by default) the
+      -- monotonic clock, and it advances /before/ threads correspondingly
+      -- wake-up from a @threadDelay@.
+      --
+      -- So we rely on the (monotonic) clock instead of using a ticker thread
+      -- and internal state, since a calling thread and the ticker thread might
+      -- wake up simultaneously from a
+      -- 'Control.Monad.Class.MonadTimer.threadDelay' call: it's ambiguous
+      -- which thread would win the race to respectively read or increment the
+      -- internal state.
```

Comment:
Not sure I understand why this matters. If the race is lost, time simply hasn't advanced yet. I guess there is an additional assumption here that that other thread that wakes up really expects to be in the new thread now?

Reply:
> If the race is lost, time simply hasn't advanced yet.

But time has advanced according to the `SystemTime`.

```
do
  threadDelay d
  t1 <- systemTimeCurrent finiteSystemTime
  s1 <- getCurrentSlot
  atomically (pure ()) -- "yield"
  t2 <- systemTimeCurrent finiteSystemTime     
  s2 <- getCurrentSlot
```

If we allow the race, then we necessarily have (b/c `io-sim`) that `t1 == t2` but it's possible that `s1 /= s2`, if the above thread had won the race against the ticker thread before yielding.

Since `hardforkBlockchainTime` (et al?) uses the `SystemTime` directly while instead other functions (eg the test infra) use `getCurrentSlot`, I was seeing confusing manifestations of the above incoherency. Thus I rewrote `getCurrentSlot` to also use `SystemTime`.

Reply:
Somewhat severely, I replaced `OracularClock.new` with `mkOracularClock`. It now uses the given `SystemTime+threadDelay` for everything -- there is no ticker thread, it's even a pure function now.